enhancements/update/object-removal-manifest-annotation.md

```diff
@@ -0,0 +1,228 @@
+---
+title: use-manifest-annotation-for-object-removal
+authors:
+  - "@jottofar"
+reviewers:
+  - "@LalatenduMohanty"
+  - "@wking"
+  - "@smarterclayton"
+  - "@deads2k"
+approvers:
+  - "@smarterclayton"
+  - "@deads2k"
+creation-date: 2020-08-03
+last-updated: 2020-09-02
+status: implementable
+---
+
+# Use Manifest Annotation For Object Removal
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Currently CVO managed object removal is handled by jobs.
+Jobs have many immutable components, so they aren't a great match for CVO managed manifests.
+This enhancement replaces that approach with a new manifest annotation requesting the CVO delete the in-cluster object instead of creating/updating it.
+This will provide a more straightforward way for engineering to remove content.
+
+## Motivation
+
+There should be a straightforward way to remove CVO managed objects.
+
+### Goals
+
+* Developers will be able to remove any of the currently managed CVO objects by modifying an existing manifest and adding the new delete annotation.
+
+### Non-Goals
+
+* This enhancement does not add any new objects for management by CVO, therefore the new delete annotation only applies to the currently supported manifests and any that may be added in the future.
+* This enhancement neither adds nor requires any application specific delete logic within CVO.
+
+## Proposal
+
+When the following annotation appears in a CVO supported manifest and is set to "true" the associated object will be removed from the cluster by the CVO.
+Values other than `true` will result in a CVO failure.
+However this should never occur in a release since any value other than `true` will result in CVO CI failure.
+```
+apiVersion: apps/v1
+...
+metadata:
+...
+  annotations:
+    release.openshift.io/delete: "true"
+```
+The existing CVO ordering scheme defined [here](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/operators.md#what-is-the-order-that-resources-get-createdupdated-in) will also be used for object removal.
+This both minimizes CVO code changes and provides a simple method of deleting multiple objects by reversing the order in which they were created.
+It is the developer's responsibility to ensure proper deletion ordering and to ensure that all items originally created by an object are deleted when that object is deleted.
+For example, an operator may have to be modified, or a new operator created, to take explicit actions to remove external resources.
+The modified or new operator would then be removed in a subsequent update.
+
+### Delete Finalization
```

Comment:
Does the CVO detect when these resources reappear?  If there's a bug somewhere, particularly for the clusteroperator resources, the resources being deleted could be recreated.  How will we know this happening and alert the cluster-admin?  The clusteroperator is the mostly likely to hit this since its creation is handled by a different operator.

Reply:
The thought on this one is that the delete is no different than the general case of having to continually reconcile the same manifest so we could alert on that general condition but it seems like an orthogonal effort.

Reply:
> The thought on this one is that the delete is no different than the general case of having to continually reconcile the same manifest so we could alert on that general condition but it seems like an orthogonal effort.

We've also had problems with the general case where the CVO constantly fights or updates various resources.  Deletion is a clear indication of intent and expectation. If that expectation is violated, a warning signal should be produced.

Reply:
> Deletion is a clear indication of intent and expectation. If that expectation is violated, a warning signal should be produced.
Agree.

Reply:
Explicitly mentioned now.